docs/specs/saml.md

```diff
@@ -436,32 +443,109 @@ The `attributes` object under items of `service_providers` is used to customize
-  - `to_saml_attribute`: Required. The `Name` of the SAML attribute to write the value.
+  - `from.user_profile.pointer`: Required. A JSON pointer to a field of the user profile. A value will be read using the JSON pointer from the user profile. And write to the SAML attribute with `Name` specified by `to.saml_attribute`.
+  - `to.saml_attribute`: Required. The `Name` of the SAML attribute to write the value.
 
-  2. A hook. It runs a hook to computes the resulting SAML attributes. See the below section [Use a hook to compute SAML attributes](#use-a-hook-to-compute-saml-attributes) for more details about the hook. The object contains the following fields:
+  2. A object uses go template to render a string as the value of a attribute. The object contains the following fields:
+
+  - `from.template`: Required. A go template which will be used to render the attribute. See the below section [Use a go template to render SAML attributes](#4_2) for more details and examples.
+  - `to.saml_attribute`: Required. The `Name` of the SAML attribute to write the value.
+
+  3. A hook. It runs a hook to computes the resulting SAML attributes. See the below section [Use a hook to compute SAML attributes](#4_3) for more details about the hook. The object contains the following fields:
 
   - `hook`: Required. The url of the hook to execute. It could be either a webhook url, or a deno hook. See [hook](./hook.md) for details.
 
   - Note that the ordering of items inside `mappings` is important. The items are run one by one, from top to bottom. Mappings item run later will override the result of the previous steps if there is a conflict, for example, written to a same attribute `Name`. So for the following example:
 
   ```yaml
   mappings:
-    - from_user_profile_attribute: /email
-      to_saml_attribute: username
-    - from_user_profile_attribute: /phone_number
-      to_saml_attribute: username
+    - from:
+        user_profile:
+          pointer: /email
+      to:
+        saml_attribute: username
+    - from:
+        user_profile:
+          pointer: /phone_number
+      to:
+        saml_attribute: username
   ```
 
   The value in the SAML attribute with `Name` equal to `username` will be the phone number of the user.
 
-### <a id="4_2"></a> Use a hook to compute SAML attributes
+### <a id="4_2"></a> Use a go template to render SAML attributes
+
+You can provide a [go template](https://pkg.go.dev/text/template) to render a SAML attribute.
+
+The data provided to the go template is the user profile object.
+
+Here is an example:
+
+Given the following `attributes` config:
+
+```yaml
+attributes:
+  definitions:
+    - name: mail
+      name_format: urn:oasis:names:tc:SAML:2.0:attrname-format:basic
+  mappings:
+    - from:
+        template: "{{.preferred_username}}@example.com"
```

Comment:
I found that there will be problems if we use html template to render the attribute.
Try this playground:
https://go.dev/play/p/3hd6eSK3d9U

You can see any email with `+` will be escaped in the attribute, which may cause problem.

So I think we need to render the template with `text/template` instead of `html/template`, with one exception that `<no value>` will be rendered as empty string.

And I suggest we modify the config to explicitly mention the type of the template:

```yaml
mappings:
  - from:
      template:
        text: "{{.email}}"
    to:
      saml_attribute: mail
```

And for any html template we support in the future, it should looks like
```yaml
template:
  html: "<p>{{.email}}</p>"
```

Reply:
```
from:
  text_template:
    template: "{{ .email }}"
  user_profile:
    pointer: /email
  html_template:
    template: "{{ .email }}"
```

```
formatGoTimeDate "2006-01-02" (parseGoTimeDate "02/01/2006" .custom_attributes.joinDate)

```

Reply:
We will modify the config to
```yaml
from:
  text_template:
    template: "{{ .email }}"
```

So that if we need a config of html template in future, we can make it as
```yaml
html_template:
  template: "{{ .email }}"
```